accounting/src/delta/delta_data_collection/tests.rs

```diff
@@ -0,0 +1,786 @@
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::*;
+
+// test function to avoid implementing Eq to compare DataDeltaUndo with DataDelta
+fn convert_undo_to_delta<T: Clone>(d: DataDeltaUndo<T>) -> DataDelta<T> {
+    match d {
+        DataDeltaUndo::Create(d) => DataDelta::Create(d),
+        DataDeltaUndo::Modify(d) => DataDelta::Modify(d),
+        DataDeltaUndo::Delete(d) => DataDelta::Delete(d),
+    }
+}
+
+#[test]
+#[rustfmt::skip]
+fn test_combine_deltas() {
+    use DataDelta::{Create, Delete, Modify};
+
+    assert_eq!(combine_deltas(&Create(Box::new('a')), Create(Box::new('b'))),                  Err(Error::DeltaDataCreatedMultipleTimes));
+    assert_eq!(combine_deltas(&Create(Box::new('a')), Modify((Box::new('a'), Box::new('b')))), Ok(DataDelta::Create(Box::new('b'))));
+    assert_eq!(combine_deltas(&Create(Box::new('a')), Delete(Box::new('a'))),                  Ok(DataDelta::Delete(Box::new('a'))));
+
+    assert_eq!(combine_deltas(&Modify((Box::new('a'), Box::new('b'))), Create(Box::new('c'))),                  Err(Error::DeltaDataCreatedMultipleTimes));
+    assert_eq!(combine_deltas(&Modify((Box::new('a'), Box::new('b'))), Modify((Box::new('c'), Box::new('d')))), Ok(DataDelta::Modify((Box::new('a'), Box::new('d')))));
+    assert_eq!(combine_deltas(&Modify((Box::new('a'), Box::new('b'))), Delete(Box::new('c'))),                  Ok(DataDelta::Delete(Box::new('c'))));
+
+    assert_eq!(combine_deltas(&Delete(Box::new('a')), Create(Box::new('b'))),                  Ok(DataDelta::Create(Box::new('b'))));
+    assert_eq!(combine_deltas(&Delete(Box::new('a')), Modify((Box::new('b'), Box::new('c')))), Err(Error::DeltaDataModifyAfterDelete));
+    assert_eq!(combine_deltas(&Delete(Box::new('a')), Delete(Box::new('b'))),                  Err(Error::DeltaDataDeletedMultipleTimes));
+}
+
+#[test]
+#[rustfmt::skip]
+fn test_combine_delta_with_undo() {
+    let delta_create = |c| DataDelta::Create(Box::new(c));
+    let delta_modify = |c1, c2| DataDelta::Modify((Box::new(c1), Box::new(c2)));
+    let delta_delete= |c| DataDelta::Delete(Box::new(c));
+
+    let undo_create = |c| DataDeltaUndo::Create(Box::new(c));
+    let undo_modify = |c1,c2| DataDeltaUndo::Modify((Box::new(c1), Box::new(c2)));
+    let undo_delete= |c| DataDeltaUndo::Delete(Box::new(c));
+
+    assert_eq!(combine_delta_with_undo(&delta_create('a'),      undo_create('b')),      Ok(undo_create('b')));
```

Comment:
I have a feeling this might be incorrect. If I create then undo, the result is a no-op, or "termination" of the operation. In other words, if block at height N has a create operation, undoing N shouldn't result in an undo operation, but in a "no-operation", because there's nothing left to do... the undo has been done.

Maybe there should be tests with multiple steps that show that this preserves its properties. Something like:

A -> (B -> C)
(A -> B) -> C

where:

- A is a no-op
- B is a create 
- C is an undo of B

what we wanna ensure is that no matter how we do it, the final result is a no-op once we reach A and anything before A.

Reply:
Maybe you are right, indeed some cases might be converted to no-op instead of propagating further. I'll try out